### PR TITLE
Implementing dataset usage with kagglehub

### PR DESCRIPTION
**Pull request by jeward414**
Implementing the usage of datasets with kagglehub.

Will allow users to use kagglehub.dataset_download("thegoanpanda/imdb-top-2000-movies").

[b/313706281](https://b.corp.google.com/issues/313706281)

Note: So far this PR adds the download functionality for datasets using kagglehub.

### CODE REVIEW

**Comment by neshdev** [comment]
Please add tests if possible.

File: src/kagglehub/handle.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 NUM_VERSIONED_MODEL_PARTS = 5  # e.g.: <owner>/<model>/<framework>/<variation>/<version>
 NUM_UNVERSIONED_MODEL_PARTS = 4  # e.g.: <owner>/<model>/<framework>/<variation>
 
+NUM_DATASET_PARTS = 2  # e.g.: <owner>/<datasetname>
+
 # TODO(b/313706281): Implement a DatasetHandle class & parse_dataset_handle method.
```
Comment (jplotts) [inline src/kagglehub/handle.py:14]: This comment should probably be removed.

File: src/kagglehub/handle.py [hunk 2]
```diff
@@ -9,6 +9,8 @@
 NUM_VERSIONED_MODEL_PARTS = 5  # e.g.: <owner>/<model>/<framework>/<variation>/<version>
 NUM_UNVERSIONED_MODEL_PARTS = 4  # e.g.: <owner>/<model>/<framework>/<variation>
 
+NUM_DATASET_PARTS = 2  # e.g.: <owner>/<datasetname>
```
Comment (jplotts) [inline src/kagglehub/handle.py:12]: There are versioned and unversioned datasets urls as well.  This probably needs to follow the same pattern as the `MODEL_PARTS` above.

Also, tiny comment suggestion for consistency/clarity:

```suggestion
NUM_DATASET_PARTS = 2  # e.g.: <owner>/<dataset>
```

File: src/kagglehub/handle.py [hunk 3]
```diff
@@ -47,7 +49,7 @@ def to_url(self) -> str:
 
 
 @dataclass
-class DatasetHandle:
+class DatasetHandle(ResourceHandle):
     owner: str
     dataset_name: str
```
Comment (jplotts) [inline src/kagglehub/handle.py:54]: tiny nit - I know this isn't added code, but since it's early, can we please change this to `dataset` vs `dataset_name`.  A couple reasons:
* Name is somewhat redundant / doesn't add information
* Consistency with models
* `_name` usually refers to something human readable like "My Cool Dataset", but this is actually a slug like "my-cool-dataset".

File: src/kagglehub/datasets_helpers.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    api_client.post("/datasets/create/new", data)
+    logger.info(
+        f"Your dataset instance has been created.\nFiles are being processed...\nSee at: {dataset_handle.to_url()}"
+    )
+
+
+def _create_dataset_instance_version(dataset_handle: DatasetHandle, files: List[str], version_notes: str = "") -> None:
+    data = {"versionNotes": version_notes, "files": [{"token": file_token} for file_token in files]}
+    api_client = KaggleApiV1Client()
+    api_client.post(f"/datasets/{dataset_handle.owner}/{dataset_handle.dataset_name}/create/", data)
```
Comment (jplotts) [inline src/kagglehub/datasets_helpers.py:31]: I don't see where this URL is supported, only `/datasets/create/version/owner/slug`.  Where is this one registered?

File: src/kagglehub/datasets_helpers.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def _create_dataset(owner_slug: str, dataset_slug: str) -> None:
+    data = {"ownerSlug": owner_slug, "slug": dataset_slug, "title": dataset_slug, "isPrivate": True}
+    api_client = KaggleApiV1Client()
+    api_client.post("/datasets/create/new", data)
+    logger.info(f"Dataset '{dataset_slug}' Created.")
+
+
+def _create_dataset_instance(dataset_handle: DatasetHandle, files: List[str]) -> None:
```
Comment (jplotts) [inline src/kagglehub/datasets_helpers.py:19]: Datasets don't have `DatasetInstance` unlike models.  I'm not sure how this would be used.  This also seems quite similar to the request above, and the `owner` and `dataset` slug aren't present, so I don't know how it would figure out which dataset to append the files to?

File: src/kagglehub/datasets_helpers.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    try:
+        api_client = KaggleApiV1Client()
+        api_client.get(f"/datasets/{owner_slug}/{dataset_slug}/get")
+    except KaggleApiHTTPError as e:
+        if e.response is not None and (
+            e.response.status_code == HTTPStatus.NOT_FOUND  # noqa: PLR1714
+            or e.response.status_code == HTTPStatus.FORBIDDEN
+        ):
+            logger.info(
+                f"Model '{dataset_slug}' does not exist or access is forbidden for user '{owner_slug}'. Creating or handling Dataset..."  # noqa: E501
```
Comment (jplotts) [inline src/kagglehub/datasets_helpers.py:63]:
```suggestion
                f"Dataset '{dataset_slug}' does not exist or access is forbidden for user '{owner_slug}'. Creating or handling Dataset..."  # noqa: E501
```

File: src/kagglehub/http_resolver.py [hunk 7]
```diff
@@ -87,3 +137,7 @@ def _build_get_instance_url_path(h: ModelHandle) -> str:
 
 def _build_download_url_path(h: ModelHandle) -> str:
     return f"models/{h.owner}/{h.model}/{h.framework}/{h.variation}/{h.version}/download"
+
+
+def _build_dataset_download_url_path(h: DatasetHandle) -> str:
+    return f"datasets/{h.owner}/{h.dataset_name}/download"
```
Comment (jplotts) [inline src/kagglehub/http_resolver.py:143]: The models and datasets APIs don't always expose the same endpoints, and this doesn't look quite right to me either.  Take a look at `dataset_api_service.proto` for more info.

**Review by jplotts — COMMENTED** [review]
Nice start, but there are a fair number of issues where the datasets and models code need to behave differently.  Some integration tests would help figure those things out.  If you're struggling with the datasets data model or endpoints, @jmasukawa should be available to consult.

File: src/kagglehub/datasets_helpers.py (hunk 4 above)
Comment (jeward414) [reply src/kagglehub/datasets_helpers.py:31]: I used the format of models as a placeholder and forgot to update this url

File: src/kagglehub/http_resolver.py [hunk 8]
```diff
@@ -11,14 +11,64 @@
...
+        url_path = _build_dataset_download_url_path(h)
+        out_path = get_cached_path(h, path)
+
+        # Create the intermediary directories
+        if path:
+            # Downloading a single file.
+            os.makedirs(os.path.dirname(out_path), exist_ok=True)
+            api_client.download_file(url_path + "/" + path, out_path, h)
+        else:
+            # Downloading the full archived bundle.
```
Comment (jmasukawa) [inline src/kagglehub/http_resolver.py:45]: @jeward414 @jplotts (cc: @rosbo )

FYI: Vincent called out an edgecase issue with unversioned handles and a new dataset version that is processed, but without a zip generated yet.

https://docs.google.com/document/d/1eNmztP3xcTit_1m_LkWi5wFt7Z4mL-00lPAs-IQTtxw/edit?usp=drivesdk&resourcekey=0-sr7en0pAcVJaSljXfhNjQg&disco=AAABJp2zl44

tldr is that we will throw an exception if a version is still zipping. for some datasets with many files or large files, this can take up to 24 hours to zip.

I offered some ideas to address in that thread, but curious your thoughts.

Comment (jeward414) [reply src/kagglehub/http_resolver.py:45]: I think the idea of throwing an exception is good, instead of letting the user not getting the latest version when they didn't specify.

File: src/kagglehub/datasets.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    Args:
+        handle: (string) the dataset handle
+        path: (string) Optional path to a file within the dataset
+        force_download: (bool) Optional flag to force download a dataset, even if it's cached.
+    Returns:
+        A string requesting the path to the requested dataset files.
+    """
+
+    h = parse_dataset_handle(handle)
+    logger.info(f"Downloading Datset: {h.to_url()} ...", extra={**EXTRA_CONSOLE_BLOCK})
```
Comment (rosbo) [inline src/kagglehub/datasets.py:24]: let's remove this log statement and let the each resolvers handle the logging. For instance, if you are in a Kaggle environment, we will "attach" the dataset and not "download" it.

**Review by rosbo — COMMENTED** [review]
Let's make sure to add integration tests here: https://github.com/Kaggle/kagglehub/tree/main/integration_tests

File: src/kagglehub/cache.py
Comment (rosbo) [inline src/kagglehub/cache.py:1]: let's make sure to add unit test to `test_cache.py`.

File: tests/test_handle.py [hunk 10]
```diff
@@ -33,3 +33,14 @@ def test_invalid_model_handle(self) -> None:
     def test_invalid_version_model_handle(self) -> None:
         with self.assertRaises(ValueError):
             parse_model_handle("google/bert/tensorFlow2/answer-equivalence-bem/invalid-version-number")
+
+    def test_dataset_handle(self) -> None:
```
Comment (rosbo) [inline tests/test_handle.py:37]: Add tests for versioned and unversioned dataset handles and name the tests accordingly.

File: tests/data/CLEANED_Spotify_artist_info_Mnth-Lstnrs.csv
Comment (rosbo) [inline tests/data/CLEANED_Spotify_artist_info_Mnth-Lstnrs.csv:1]: Remove this file?

File: tests/test_dataset_upload.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+from .utils import create_test_http_server
+
+GET_DATASET = "/datasets/view/akankshaaa013/top-grossing-movies-dataset"
+CREATE_DATASET = "/datasets/create/new"
+
+TEMP_TEST_FILE = "temp-test-file"
+
+
+
+class KaggleAPIHandler(BaseHTTPRequestHandler):
```
Comment (rosbo) [inline tests/test_dataset_upload.py:20]: We are now using flask. This was the old style of writing test which relied on the base http package. Make sure to follow the new style. 

I see you added some `dataset_upload_stub.py`. Maybe you are actively working on that. If that's the case, you can disregard this comment.

**Review by rosbo — COMMENTED** [review]
General comment. For future PRs/CLs, try to follow: [go/small-cls](https://goto.google.com/small-cls)

In this particular case, a natural split would have been to create two PRs. The first for download and the second for upload.

File: src/kagglehub/datasets.py (hunk 9 above)
Comment (rosbo) [reply src/kagglehub/datasets.py]: Sorry, looks like Nesh added it for models and added an extra to exclude it from the console log (only logs to file). We can keep it to match models.

Make sure to fix the typo: "s/Datsaet/Dataset"

File: tests/server_stubs/dataset_download_stub.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+# See https://cloud.google.com/storage/docs/xml-api/reference-headers#xgooghash
+GCS_HASH_HEADER = "x-goog-hash"
+
+@app.route("/", methods=["HEAD"])
+def head() -> ResponseReturnValue:
+    return "", 200
+
+
+@app.route("/api/v1/datsets/view/<owner_slug>/<dataset_slug>", methods=["GET"])
```
Comment (rosbo) [inline tests/server_stubs/dataset_download_stub.py:21]: typo: datsets -> datasets.

File: src/kagglehub/cache.py [hunk 13]
```diff
@@ -37,6 +41,8 @@ def get_cached_path(handle: ResourceHandle, path: Optional[str] = None) -> str:
 def get_cached_archive_path(handle: ResourceHandle) -> str:
     if isinstance(handle, ModelHandle):
         return _get_model_archive_path(handle)
+    elif isinstance(handle, DatasetHandle):
+        return _get_dataset_path(handle)
```
Comment (rosbo) [inline src/kagglehub/cache.py:45]: The issue is here. It should return the archive path with the archive filename.

Currently, it downloads the archive file and name it as the directory.

```
name = '/tmp/tmp2_r6k2ur/datasets/sarahjeffreson/featured-spotify-artiststracks-with-metadata/2'
```

This creates a `2` file.

then, later, you try to create a directory with the same path and it fails.

Comment (rosbo) [reply src/kagglehub/cache.py:45]: See how we create an archive path at line #43.

**Review by rosbo — COMMENTED** [review]
Also, another major part is to add support for the KaggleCacheResolver: https://github.com/Kaggle/kagglehub/blob/main/src/kagglehub/kaggle_cache_resolver.py

This is the part that attach a dataset when running inside a notebook.
